File: csharp/src/0209.cs
using System;
using System.Diagnostics;

public class Solution {

    public int MinSubArrayLen(int target, int[] nums) {
        int n = nums.Length;
        int m = int.MaxValue;
        int s = 0;
        int i = 0;
        int j = 0;
        while (j < n) {
            s += nums[j++];
            while (s >= target) {
                s -= nums[i++];
                m = Math.Min(m, j - i + 1);
            }
        }
        return m == int.MaxValue ? 0 : m;
    }

    static void Main(string[] args) {
        var o = new Solution();

        Debug.Assert(o.MinSubArrayLen(7, new[] { 2, 3, 1, 2, 4, 3 }) == 2);
        Debug.Assert(o.MinSubArrayLen(4, new[] { 1, 4, 4 }) == 1);
        Debug.Assert(o.MinSubArrayLen(11, new[] { 1, 1, 1, 1, 1, 1, 1, 1 }) == 0);

        var timer = new Stopwatch();
        timer.Start();
        for (var i = 0; i < 1000000; i++) {
            o.MinSubArrayLen(7, new[] { 2, 3, 1, 2, 4, 3 });
            o.MinSubArrayLen(4, new[] { 1, 4, 4 });
            o.MinSubArrayLen(11, new[] { 1, 1, 1, 1, 1, 1, 1, 1 });
        }
        timer.Stop();
        Console.WriteLine(timer.Elapsed.TotalMilliseconds);
    }
}

File: cpp/src/0209.cpp
/*
[209. Minimum Size Subarray Sum (Medium)](https://leetcode.com/problems/minimum-size-subarray-sum/)

@topic: Array
@topic: Binary Search
@topic: Sliding Window
@topic: Prefix Sum

Constraints:
  - `1 <= target <= 10^9`
  - `1 <= nums.length <= 10^5`
  - `1 <= nums[i] <= 10^4`
*/

#include <vector>

using std::min;
using std::vector;

class Solution {
   public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int res = INT_MAX;
        int sum = 0;
        int slow = 0;
        int fast = 0;
        int size = nums.size();
        while (fast < size) {
            sum += nums[fast++];
            while (sum >= target) {
                sum -= nums[slow++];
                res = min(res, fast - slow + 1);
            }
        }
        return res == INT_MAX ? 0 : res;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>

using std::tuple;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    Solution o;

    vector<tuple<int, vector<int>, int>> CASES = {
        {7, {2, 3, 1, 2, 4, 3}, 2},
        {4, {1, 4, 4}, 1},
        {11, {1, 1, 1, 1, 1, 1, 1, 1}, 0},
    };

    for (auto& [target, nums, excepted] : CASES) {
        assert(o.minSubArrayLen(target, nums) == excepted);
    }

    auto start = system_clock::now();
    for (auto& [target, nums, _] : CASES) {
        for (int i = 0; i < 100000; ++i) {
            o.minSubArrayLen(target, nums);
        }
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}
